File: Python3/valid_perfect_square.py
class Solution:
    def isPerfectSquare(self, num: int) -> bool:
        if num < 2: return True;
        
        x = num // 2
        t = x * x
        
        while t > num:
            x //= 2
            t = x * x
        
        for i in range(x, 2 * x + 1):
            if i * i == num: return True
        
        return False

File: C++/valid_perfect_square.cpp
class Solution {
public:
    bool isPerfectSquare(int num) {
        if (num < 2) return true;
        
        long x = num / 2, t = x * x;
        
        while (t > num) {
            x /= 2;
            t = x * x;
        }
        
        for (long i = x; i <= 2 * x; ++i) {
            if (i * i == num) return true;
        }
        
        return false;
    }
};
